src/Test/Utilities/Portable/Metadata/MetadataReaderUtils.cs

```diff
@@ -265,9 +265,147 @@ public static string Dump(this MetadataReader reader, EntityHandle handle)
+                    var signature =  decoder.DecodeMethodSignature(ref blob);
+                    var parameters = signature.ParameterTypes.Join(", ");
+                    return $"MethodDef: {signature.ReturnType} {reader.GetString(method.Name)}({parameters})";
                 default:
                     return handle.Kind.ToString();
             }
         }
+
+        private sealed class ConstantSignatureVisualizer : ISignatureTypeProvider<string, object>
+        {
+            public static readonly ConstantSignatureVisualizer Instance = new ConstantSignatureVisualizer();
+
+            public string GetArrayType(string elementType, ArrayShape shape)
+            {
+                return elementType + "[" + new string(',', shape.Rank) + "]";
+            }
+
+            public string GetByReferenceType(string elementType)
+            {
+                return elementType + "&";
+            }
+
+            public string GetFunctionPointerType(MethodSignature<string> signature)
+            {
+                return "method-ptr";
+            }
+
+            public string GetGenericInstantiation(string genericType, ImmutableArray<string> typeArguments)
+            {
+                return genericType + "{" + string.Join(", ", typeArguments) + "}";
+            }
+
+            public string GetGenericMethodParameter(object genericContext, int index)
+            {
+                return "!!" + index;
+            }
+
+            public string GetGenericTypeParameter(object genericContext, int index)
+            {
+                return "!" + index;
+            }
+
+            public string GetModifiedType(string modifier, string unmodifiedType, bool isRequired)
+            {
+                return (isRequired ? "modreq" : "modopt") + "(" + modifier + ") " + unmodifiedType;
+            }
+
+            public string GetPinnedType(string elementType)
+            {
+                return "pinned " + elementType;
+            }
+
+            public string GetPointerType(string elementType)
+            {
+                return elementType + "*";
+            }
+
+            public string GetPrimitiveType(PrimitiveTypeCode typeCode)
+            {
+                return typeCode.ToString();
+            }
+
+            public string GetSZArrayType(string elementType)
+            {
+                return elementType + "[]";
+            }
+
+            public string GetTypeFromDefinition(MetadataReader reader, TypeDefinitionHandle handle, byte rawTypeKind)
+            {
+                var typeDef = reader.GetTypeDefinition(handle);
+                var name = reader.GetString(typeDef.Name);
+                return typeDef.Namespace.IsNil ? name : reader.GetString(typeDef.Namespace) + "." + name;
+            }
+
+            public string GetTypeFromReference(MetadataReader reader, TypeReferenceHandle handle, byte rawTypeKind)
+            {
+                var typeRef = reader.GetTypeReference(handle);
```

Comment:
You can steal some helper from 
[https://github.com/dotnet/symreader-converter/blob/master/src/Microsoft.DiaSymReader.Converter/PdbConverterPortableToWindows.MetadataModel.cs#L172](https://github.com/dotnet/symreader-converter/blob/master/src/Microsoft.DiaSymReader.Converter/PdbConverterPortableToWindows.MetadataModel.cs#L172)

---
In reply to: [107314698](https://github.com/dotnet/roslyn/pull/17558#discussion_r107314698) [](ancestors = 107314698)